syntax = "proto3";
package redis_request;

import "protobuf/routes.proto";

enum RequestType {
    /// Invalid request type
    InvalidRequest = 0;
    /// An unknown command, where all arguments are defined by the user.
    CustomCommand = 1;
    /// Type of a get string request.
    GetString = 2;
    /// Type of a set string request.
    SetString = 3;
}

message Command {
    message ArgsArray {
        repeated string args = 1;
    }

    RequestType request_type = 1;
    oneof args {
        ArgsArray args_array = 2;
        uint64 args_vec_pointer = 3;
    }
}

message Transaction {
    repeated Command commands = 1;
}

message RedisRequest {
    uint32 callback_idx = 1;
    
    oneof command {
        Command single_command = 2;
        Transaction transaction = 3;
    }
    routes.Routes route = 4;
}
